blast/ncbi/doc/man/fmerge.1
.TH FMERGE 1 2002-08-23 NCBI "NCBI Tools User's Manual"
.SH NAME
fmerge \- merge FASTA databases
.SH SYNOPSIS
.B fmerge
[\|\fB\-\fP\|]
[\|\fB\-i\fP\ \fIfilename\fP\|]
[\|\fB\-l\fP\ \fIfilename\fP\|]
[\|\fB\-m\fP\ \fIfilename\fP\|]
[\|\fB\-n\fP\ \fIfilename\fP\|]
\fB\-t\fP\ \fIN\fP
.SH DESCRIPTION
\fBfmerge\fP is a program to merge FASTA databases.
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\fP
Print usage message
.TP
\fB\-i\fP\ \fIfilename\fP
File for index storage (default is \fBindex.nr\fP)
.TP
\fB\-l\fP\ \fIfilename\fP
Log file name (default is \fBfmerge.log\fP)
.TP
\fB\-m\fP\ \fIfilename\fP
Read new sequences (in FASTA format) from \fIfilename\fP (default is
\fBmonth.aa\fP)
.TP
\fB\-n\fP\ \fIfilename\fP
Update \fIfilename\fP (default is \fBnr\fP)
.TP
\fB\-t\fP\ \fIN\fP
Job mode:
.RS
.PD 0
.IP 1
Create index only
.IP 2
Update database from existing index
.IP 3
Create index and update database
.PD
.RE
.SH AUTHOR
The National Center for Biotechnology Information.


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

blast/ncbi/demo/fmerge.c
/*****************************************************************************

    Name: fmerge.c                               

    Description: Program for merging FASTA databases.

    Author: Sergei Shavirin

   ***************************************************************************

                          PUBLIC DOMAIN NOTICE
              National Center for Biotechnology Information

    This software/database is a "United States Government Work" under the
    terms of the United States Copyright Act.  It was written as part of    
    the author's official duties as a United States Government employee
    and thus cannot be copyrighted.  This software/database is freely
    available to the public for use. The National Library of Medicine and
    the U.S. Government have not placed any restriction on its use or
    reproduction.

    Although all reasonable efforts have been taken to ensure the accuracy
    and reliability of the software and data, the NLM and the U.S.
    Government do not and cannot warrant the performance or results that
    may be obtained by using this software or data. The NLM and the U.S.
    Government disclaim all warranties, express or implied, including
    warranties of performance, merchantability or fitness for any
    particular purpose.

    Please cite the author in any work or product based on this material.

   ***************************************************************************

    Modification History:
           20 August of 1996 - Shavirin  -  originaly written

    Bugs and restriction on use:

    Notes:

*****************************************************************************/
#include <ncbi.h>

#define INCLENGTH 256
#define STRLENGTH 128

typedef struct MDeflineID {
  Uint4   gi;
  CharPtr defline;
  struct MDeflineID *next;
} MDeflineID, PNTR MDeflineIDPtr;

typedef struct FMergeID {
  MDeflineIDPtr id;
  CharPtr sequence;
} FMergeID, PNTR FMergeIDPtr;

Uint4Ptr GetIndexFromFasta(FILE *fd, Uint4Ptr MaxIndexCount);
Uint4Ptr GetIndexFromIndex(FILE *fd, Uint4Ptr MaxIndexCount);

static  FMergeIDPtr NextFastaFromFasta(FILE *fd);
Boolean IfGiInIndex(Uint4Ptr index, Uint4 MaxIndexCount, Uint4 gi);
Int4 MergeFMergeID(FMergeIDPtr fasta, 
                   FILE *fd, FILE *fd_ind,
                   Uint4Ptr index, 
                   Uint4 MaxIndexCount);
void FMergeIDFree(FMergeIDPtr fasta);

#define NUMARG 5

Args dump_args[NUMARG] = {
  {"Job mode (this value must be set):\n"
   "         1 - create index only\n"
   "         2 - update database from existing index\n"
   "         3 - create index and update database\n"
   "                                   ",

   NULL, NULL,NULL,FALSE,'t',ARG_STRING, 0.0,0,NULL},
  {"Logfile name:","fmerge.log",
     NULL,NULL,TRUE,'l',ARG_FILE_OUT, 0.0,0,NULL},
  {"Input file for indexing\n"
   "         Name of FASTA file, that you are going to update\n"
   "                                   ",
   "nr", NULL,NULL,TRUE,'n',ARG_FILE_IN,0.0,0,NULL},

  {"Input file for new gis:\n"
   "         Name of FASTA file with new sequences\n"
   "                                   ", 
   "month.aa", NULL,NULL,TRUE,'m',ARG_FILE_IN,0.0,0,NULL},
  {"Index file name\n"
   "         Name of file for index storage\n"
   "                                   ",
   "index.nr", NULL,NULL,TRUE,'i',ARG_FILE_IN,0.0,0,NULL},
};

#define JobModeStr  (const char *) dump_args[0].strvalue 
#define LogFileName (const char *) dump_args[1].strvalue  
#define NRFile      (const char *) dump_args[2].strvalue
#define MonthFile   (const char *) dump_args[3].strvalue
#define IndexFile   (const char *) dump_args[4].strvalue

#define INDEX_MODE  1
#define UPDATE_MODE 2
#define FULL_MODE   3

/* ------------------------------------------------------------------
                This is handler for HeapSort function
   ------------------------------------------------------------------*/
static int LIBCALLBACK intcompare(VoidPtr i, VoidPtr j)
{
  if (*(int *)i > *(int *)j)
    return (1);
  if (*(int *)i < *(int *)j)
    return (-1);
  return (0);
}

Int2 Main(void) 
{
  Uint4Ptr index = NULL;
  FILE *fd, *fd_ind, *fdout;
  Uint4 i, MaxIndexCount;
  FMergeIDPtr fasta;
  Int4 Added =0, TotalAdded =0, SequenceAdded =0;
  Int4 JobMode;

  if ( !GetArgs ("fmerge",NUMARG,dump_args) ) {
    return 1;
  }
  if ( !ErrSetLog (LogFileName) ) {
    ErrShow();
  } else {
    ErrSetOpts (ERR_CONTINUE, ERR_LOG_ON);
  }

  JobMode = atoi(JobModeStr);

  switch(JobMode) {
  case INDEX_MODE:
    ErrLogPrintf("fmerge started in \"Create index\" mode\n");
    break;
  case UPDATE_MODE:
    ErrLogPrintf("fmerge started in \"Update database\" mode\n");
    break;
  case FULL_MODE:
    ErrLogPrintf("fmerge started in \"Full mode\" mode\n");
    break;
  default:
    printf("fmerge mode \"%s\" invalid, exiting...\n", 
           JobModeStr);
    exit(1);
  }

  /* First creating index if necessary */

  if(JobMode == INDEX_MODE || JobMode == FULL_MODE) {

    if((fd = FileOpen(NRFile, "r")) == NULL) {
      ErrLogPrintf("Cannot open input file. Indexing failed...\n");
      exit(1);
    }
    ErrLogPrintf("Using file \"%s\" to create index\n", NRFile);
    
    if((index = GetIndexFromFasta(fd, &MaxIndexCount)) == NULL) {
      ErrLogPrintf("Error in creating gi index from FASTA file\n");
    }
    FileClose(fd);
    ErrLogPrintf("Created index of %d gis\n", MaxIndexCount);
    
    fd_ind = FileOpen(IndexFile, "w");
    for(i = 0; i < MaxIndexCount; i++)
      fprintf(fd_ind, "%ld\n", index[i]); 
    FileClose(fd_ind);
	
    if(JobMode == INDEX_MODE) {
      MemFree(index);
      exit(0);
    }
  }

  if(JobMode == UPDATE_MODE) {
    
    if((fd_ind = FileOpen(IndexFile, "r")) == NULL) {
      ErrLogPrintf("Cannot open file with index. Exiting...\n");
      exit(1);
    }
    if((index = GetIndexFromIndex(fd_ind, &MaxIndexCount)) == NULL) {
      ErrLogPrintf("Error in creating gi index from FASTA file\n");
    }
    FileClose(fd_ind);
    ErrLogPrintf("Created index of %d gis\n", MaxIndexCount);
  }

  /* Now searching FASTA file with new gis*/

  if((fd = FileOpen(MonthFile, "r")) == NULL) {
    ErrLogPrintf("Cannot open input file. Indexing failed...\n");
    exit(1);
  }
  ErrLogPrintf("Using file \"%s\" as FASTA database with new gis\n", 
               MonthFile);

  if((fdout = FileOpen(NRFile, "a+")) == NULL) {
    ErrLogPrintf("Error opening %s file\n", NRFile);
    exit(1);
  }
  
  if((fd_ind = FileOpen(IndexFile, "a+")) == NULL) {
    ErrLogPrintf("Error opening %s file\n", IndexFile);
    exit(1);
  }

  while((fasta = NextFastaFromFasta(fd)) != NULL) {
    if((Added = MergeFMergeID(fasta, fdout, fd_ind, 
                              index, MaxIndexCount)) != 0) {
      TotalAdded += Added;
      SequenceAdded++;
    }
  }
  MemFree(index);
  FileClose(fd);
  FileClose(fd_ind);
  ErrLogPrintf("Total gis added:       %d\n", TotalAdded);
  ErrLogPrintf("Total sequences added: %d\n", SequenceAdded);
  return 0;
}
Uint4Ptr GetIndexFromIndex(FILE *fd, Uint4Ptr MaxIndexCount)
{
 Uint4Ptr index;
 Uint4 IndexSize = INCLENGTH;
 Uint4 i =0, gi;

 index = MemNew(IndexSize*4);
 
 while (fscanf(fd, "%ld", &gi) > 0) {
   
   if(i == IndexSize) {
     IndexSize += INCLENGTH;
     index = Realloc(index, IndexSize*4);
   }
   index[i] = gi;
   i++;
 }
 *MaxIndexCount = i;
 HeapSort(index, *MaxIndexCount, sizeof(Uint4), intcompare);
 return index;
}

Uint4Ptr GetIndexFromFasta(FILE *fd, Uint4Ptr MaxIndexCount)
{
  Char str[20000];
  CharPtr pos;
  CharPtr start;
  Uint4Ptr index;
  Uint4 i =0, gi;
  Uint4 IndexSize = INCLENGTH;

  index = MemNew(IndexSize*4);

  while(fgets (str, sizeof (str), fd) != NULL) {
    start = str;
    while((pos = StringStr(start, "\1gi|")) != NULL  ||
          (pos = StringStr(start, ">gi|")) != NULL) {
      if(!sscanf(pos+4, "%ld", &gi)) {
        ErrLogPrintf("Error parsing gi number. Indexing failed...\n");
        exit(1);
      }
      /* Here we got next gi nnumber and will add it to the index */

      if(i == IndexSize) {
        IndexSize += INCLENGTH;
        index = Realloc(index, IndexSize*4);
      }
      index[i] = gi;
      i++;
      start = pos+4;
    }
  }
  *MaxIndexCount = i;
  HeapSort(index, *MaxIndexCount, sizeof(Uint4), intcompare);
  return index;
}

Boolean IfGiInIndex(Uint4Ptr index, Uint4 MaxIndexCount, Uint4 gi)
{
  Uint4 high_index, low_index, new_index; 

  if(MaxIndexCount == 0 || index == NULL || gi == 0)
    return FALSE;

  low_index  = 0;
  high_index = MaxIndexCount;
  new_index  = (low_index+high_index)/2;

  while(new_index != low_index && new_index != high_index){
    if (gi > index[new_index]) {
      low_index = new_index;
    } else if (gi < index[new_index]) {
      high_index = new_index;
    } else {	/* scores are equal. */
      return TRUE;
    }
    new_index = (low_index+high_index)/2;
  }
  if (gi == index[new_index]) 
    return TRUE;
  else
    return FALSE;
}

static FMergeIDPtr NextFastaFromFasta(FILE *fd)
{
  CharPtr str, start;
  Int4 i=0,j=0;
  Char ch;
  Int4 gi;
  CharPtr pos;
  Int4 f_pos;
  Char tmpbuff[2048];
  FMergeIDPtr fasta;
  MDeflineIDPtr id, last_id;
  CharPtr TmpSequence;
  Int4 SeqSize = STRLENGTH;

  if((fasta = MemNew(sizeof(FMergeID))) == NULL) {
    ErrLogPrintf("Cannot allocate memory for Fasta ID\n");
    exit(1);
  }
  fasta->id = NULL;

  str = MemNew(SeqSize + 5);
  str[0] = NULLB;
  for(i=0; (FileRead(&ch, sizeof(ch), 1, fd) == 1); i++) {      
    if((str[i] = ch) == '\n' || ch == '\r')
      break;
    if (i > SeqSize) {
      SeqSize = i + STRLENGTH;
      str = Realloc(str, SeqSize + 5);
    }
  }
  str[i+1] = NULLB;
  start = str;
  while((pos = StringStr(str, "\1gi|")) != NULL ||
        (pos = StringStr(str, ">gi|")) != NULL) {
    if(!sscanf(pos+4, "%ld", &gi)) {
      ErrLogPrintf("Error parsing gi number. Indexing failed...\n");
      exit(1);
    }
    pos++;
    for(j = 0;pos[j] != '\1' && 
              pos[j] != '\n' && 
              pos[j] != '\r'; j++) {
      tmpbuff[j] = pos[j];
    }
    tmpbuff[j] = '\0';

    id = MemNew(sizeof(MDeflineID));
    id->defline = StringSave(tmpbuff);
    id->gi = gi;
    id->next = NULL;

    if(fasta->id == NULL) {
      fasta->id = id;
      last_id = fasta->id;
    } else {
      last_id->next = id;
      last_id = last_id->next;
    }
    str = pos+4;
  }

  MemFree(start);

  if(fasta->id == NULL) {
    MemFree(fasta);
    return NULL;
  }

  TmpSequence = MemNew(SeqSize + 5);
  TmpSequence[0] = NULLB;


  while(TRUE) {
    f_pos = ftell(fd);
    if(fgets(tmpbuff, sizeof (tmpbuff), fd) == NULL)
      break;
    if(StringStr(tmpbuff, ">gi|") != NULL) { /* new fasta entry started */
      fseek(fd, (long) f_pos, SEEK_SET);
      break;
    }
    StringCat(TmpSequence, tmpbuff);
    SeqSize += STRLENGTH + 5;
    TmpSequence = Realloc(TmpSequence, SeqSize);
  }
  fasta->sequence = TmpSequence;
  return fasta;
}

Int4 MergeFMergeID(FMergeIDPtr fasta, FILE *fd, FILE *fd_ind,
                   Uint4Ptr index, Uint4 MaxIndexCount) 
{
  Boolean SomeNew = FALSE;
  MDeflineIDPtr id;
  Int4 TotalNot =0;

  id = fasta->id;
  while(id != NULL) {
    if(IfGiInIndex(index, MaxIndexCount, id->gi)) {
      id = id->next;
      continue;
    }
    TotalNot++;

    fprintf(fd_ind, "%ld\n", id->gi);

    if(SomeNew)
      fprintf(fd, "\1%s", id->defline);
    else {
      fprintf(fd, ">%s", id->defline);
      SomeNew = TRUE;
    }
    id = id->next;
  }

  if(SomeNew)
    fprintf(fd, "\n%s", fasta->sequence);
  
  FMergeIDFree(fasta);
  return TotalNot;
}
void FMergeIDFree(FMergeIDPtr fasta)
{
  MDeflineIDPtr id, id1;

  MemFree(fasta->sequence);
  id = fasta->id;
  while(id != NULL) {
    MemFree(id->defline);
    id1 = id;
    id = id->next;
    MemFree(id1);
  }
  MemFree(fasta);
}

